cachito/workers/pkg_managers/pip.py

```diff
@@ -2175,3 +2181,60 @@ def get_raw_component_name(requirement):
         raw_component_name = f"{repo_name}/{filename}"
 
     return raw_component_name
+
+
+def _iter_zip_file(file_path: Path):
+    with zipfile.ZipFile(file_path, "r") as zf:
+        return zf.namelist()
+
+
+def _iter_tar_file(file_path: Path):
+    with tarfile.open(file_path, "r") as tar:
+        for member in tar:
+            yield member.name
+
+
+def is_pkg_info_dir(path):
+    """Simply check whether a path represents the PKG_INFO directory.
+
+    Generally, it is in the format for example: pkg-1.0/PKG_INFO
+
+    :param str path: a path.
+    :return: True if it is, otherwise False is returned.
+    :rtype: bool
+    """
+    parts = os.path.split(path)
+    return len(parts) == 2 and parts[1] == "PKG-INFO"
+
+
+def check_metadata_in_sdist(sdist_path: Path):
+    """Check if a downloaded sdist package has metadata.
+
+    :param sdist_path: the path of a sdist package file.
+    :type sdist_path: pathlib.Path
+    :raise ValidationError: if the sdist cannot be validated.
+    """
+    if str(sdist_path).endswith(ZIP_FILE_EXT):
+        try:
+            files_iter = _iter_zip_file(sdist_path)
+        except zipfile.BadZipFile as e:
+            raise ValidationError(f"Cannot open {sdist_path} as a Zip file. Error: {str(e)}")
+    elif str(sdist_path).endswith(COMPRESSED_TAR_EXT):
+        log.warning("Skip checking metadata from compressed sdist %s", sdist_path.name)
+        return
+    elif any(map(str(sdist_path).endswith, SDIST_FILE_EXTENSIONS)):
+        files_iter = _iter_tar_file(sdist_path)
+    else:
+        raise ValidationError(
+            f"Cannot check metadata from {sdist_path}, "
+            f"which does not have a known supported extension.",
+        )
+
+    try:
+        if not any(map(is_pkg_info_dir, files_iter)):
+            raise ValidationError(
+                f"{sdist_path.name} does not include metadata. "
+                f"Please download from the package from git or a direct download URL."
+            )
+    except tarfile.ReadError as e:
+        raise ValidationError(f"Cannot open {sdist_path} as a Tar file. Error: {str(e)}")
```

Comment:
It's a bit odd that `_iter_zip_file()` raises `zipfile.BadZipFile` as soon as you call it but `_iter_tar_file()` raises the error when your iterate through the result. Could `_iter_tar_file()` be changed to behave the same way?

Reply:
The major consideration behind the current behavior of `_iter_tar_file()` is the comment[1]. I originally wanted the same iteration behavior for zip file, but AFAIK the ZipFile does not support it.

[1] https://github.com/pypa/warehouse/pull/5609/files#diff-bffd49e4f943217b54b4cfcf5835737bec7a145360bd1dc2ec1541d2dcafad16R587

Reply:
Ah, that makes sense. So an invalid tar file will *only* raise an error when you start to iterate it? Is it possible that `tarfile.open()` will raise the error? (that would currently be uncaught)

Reply:
Just had a quick look. The error is raised just when start the iteration, when us tarfile object as a context manager object. Calling `tarfile.open` directly will raise the error immediately.

Reply:
Actually function `_iter_tar_file` starts to execute only when the iteration happens. Due to the aforementioned ability of ZipFile, the two `_iter-*` functions returns different type of value, one is a list and another one is iterator.

Reply:
Oh, that's true! Would it make sense to move the try/except blocks into the `_iter_*` functions? That way, the caller doesn't have to worry about the list vs. generator difference

Reply:
Alternatively, you could make the zip one a generator with `yield from zf.namelist()`